external_programs/utils.py

```diff
@@ -14,39 +17,86 @@
+    # Add constant padding of samples before and after the non-silent indexes
+    beginningPadCheck = index[0] - samplePad
+    endingPadCheck = index[1] + samplePad
+    numSamples = len(signal) - 1
+
+    # Ensure that padding does not exceed the buffer
+    # We cannot have a negative starting index, so make it zero. If its positive, keep it as it is.
+    if (beginningPadCheck < 0):
+        beginningIndex = 0
+    else:
+        beginningIndex = beginningPadCheck
+
+    # We cannot have an ending index greater than the length of the buffer; In this case, return the last index of
+    # the buffer.
+    if (endingPadCheck > numSamples):
+        endingIndex = numSamples
+    else:
+        endingIndex = endingPadCheck
+
+    # Return the trimmed signal with padding to prevent slight cutoff in commands
+    return signal[beginningIndex:endingIndex]
+
+# Use this function for processing if you already have the signal loaded using librosa
+def processPreloadedAudio(inputSignal, inputSignalSampleRate):
+    highPassedInputSignal = highPassFilterStage(inputSignal, inputSignalSampleRate)
+    denoisedSignal = denoiseStage(highPassedInputSignal, inputSignalSampleRate)
+    processedSignal = trimStage(denoisedSignal)
+    return processedSignal, inputSignalSampleRate
+
 def extract_feature(file_name, **kwargs):
 
     mfcc = kwargs.get("mfcc")
     chroma = kwargs.get("chroma")
     mel = kwargs.get("mel")
     contrast = kwargs.get("contrast")
     tonnetz = kwargs.get("tonnetz")
-    with soundfile.SoundFile(file_name) as sound_file:
-        X = sound_file.read(dtype="float32")
-        sample_rate = sound_file.samplerate
-        if chroma or contrast:
-            stft = np.abs(librosa.stft(X))
-        result = np.array([])
-        if mfcc:
-            mfccs = np.mean(librosa.feature.mfcc(
-                y=X, sr=sample_rate, n_mfcc=40).T, axis=0)
-            result = np.hstack((result, mfccs))
-        if chroma:
-            chroma = np.mean(librosa.feature.chroma_stft(
-                S=stft, sr=sample_rate).T, axis=0)
-            result = np.hstack((result, chroma))
-        if mel:
-            mel = np.mean(librosa.feature.melspectrogram(
-                X, sr=sample_rate).T, axis=0)
-            result = np.hstack((result, mel))
-        if contrast:
-            contrast = np.mean(librosa.feature.spectral_contrast(
-                S=stft, sr=sample_rate).T, axis=0)
-            result = np.hstack((result, contrast))
-        if tonnetz:
-            tonnetz = np.mean(librosa.feature.tonnetz(
-                y=librosa.effects.harmonic(X), sr=sample_rate).T, axis=0)
-            result = np.hstack((result, tonnetz))
+
+    inputSignal, inputSignalSampleRate = librosa.load(file_name, sr=None)
+
+    #audio processing stages invoked here
+    X, sample_rate = processPreloadedAudio(inputSignal, inputSignalSampleRate)
+
+    if chroma or contrast:
+        stft = np.abs(librosa.stft(X))
+    result = np.array([])
```

Comment:
Is this indentation an issue?

Reply:
I don't believe so; the indentation matches the original code:
![unknown](https://user-images.githubusercontent.com/46769411/135680776-6a0e33f0-0670-4a69-a7eb-4789b7593226.png)

The with block is much easier to read, but the librosa load function returns two values (e.g., signal and the sample rate) that have to be saved, so i omitted it.